src/coreclr/jit/gentree.cpp

```diff
@@ -32593,6 +32593,76 @@ GenTree* Compiler::gtFoldExprHWIntrinsic(GenTreeHWIntrinsic* tree)
         {
             switch (ni)
             {
+#if defined(TARGET_ARM64)
+                case NI_Vector64_ExtractMostSignificantBits:
+#elif defined(TARGET_XARCH)
+                case NI_Vector256_ExtractMostSignificantBits:
+                case NI_X86Base_MoveMask:
+                case NI_AVX_MoveMask:
+                case NI_AVX2_MoveMask:
+#endif
+                case NI_Vector128_ExtractMostSignificantBits:
+                {
+                    simdmask_t simdMaskVal;
+
+                    switch (simdSize)
+                    {
+                        case 8:
+                        {
+                            EvaluateExtractMSB<simd8_t>(simdBaseType, &simdMaskVal, cnsNode->AsVecCon()->gtSimd8Val);
+                            break;
+                        }
+
+                        case 16:
+                        {
+                            EvaluateExtractMSB<simd16_t>(simdBaseType, &simdMaskVal, cnsNode->AsVecCon()->gtSimd16Val);
+                            break;
+                        }
+
+#if defined(TARGET_XARCH)
+                        case 32:
+                        {
+                            EvaluateExtractMSB<simd32_t>(simdBaseType, &simdMaskVal, cnsNode->AsVecCon()->gtSimd32Val);
+                            break;
+                        }
+
+                        case 64:
+                        {
+                            EvaluateExtractMSB<simd64_t>(simdBaseType, &simdMaskVal, cnsNode->AsVecCon()->gtSimd64Val);
+                            break;
+                        }
+#endif // TARGET_XARCH
+
+                        default:
+                        {
+                            unreached();
+                        }
+                    }
+
+                    uint64_t mask;
+                    memcpy(&mask, &simdMaskVal.u64[0], sizeof(uint64_t));
```

Comment:
could it be just `mask = simdMaskVal.u64[0];` or am I missing something 🙂

Reply:
It's a union and so UB to do so without checking the "tag". Checking the tag would require handling each of the other union fields and we want to avoid that.

All 3 compilers do the "better" thing for architectures where that's safe to do.